rclcpp/src/rclcpp/client.cpp

```diff
@@ -42,3 +47,58 @@ ClientBase::get_client_handle() const
 {
   return &client_handle_;
 }
+
+bool
+ClientBase::service_is_ready() const
+{
+  bool is_ready;
+  rcl_ret_t ret =
+    rcl_service_server_is_available(this->get_rcl_node_handle(), &client_handle_, &is_ready);
+  if (ret != RCL_RET_OK) {
+    throw_from_rcl_error(ret, "rcl_service_server_is_available failed");
+  }
+  return is_ready;
+}
+
+bool
+ClientBase::wait_for_service_nanoseconds(std::chrono::nanoseconds timeout)
+{
+  auto start = std::chrono::steady_clock::now();
+  // check to see if the server is ready immediately
+  if (this->service_is_ready()) {
+    return true;
+  }
+  if (timeout == std::chrono::nanoseconds(0)) {
+    // check was non-blocking, return immediately
+    return false;
+  }
+  // make an event to reuse, rather than create a new one each time
+  auto event = node_->get_graph_event();
+  // update the time even on the first loop to account for time in first server_is_read()
+  std::chrono::nanoseconds time_to_wait = timeout - (std::chrono::steady_clock::now() - start);
+  if (timeout > std::chrono::nanoseconds(0) && time_to_wait < std::chrono::nanoseconds(0)) {
+    // Do not allow the time_to_wait to become negative when timeout was originally positive.
+    // Setting time_to_wait to 0 will allow one non-blocking wait because of the do-while.
+    time_to_wait = std::chrono::nanoseconds(0);
+  }
+  // continue forever if timeout is negative, otherwise continue until out of time_to_wait
+  // *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
+  do {
+    node_->wait_for_graph_change(event, time_to_wait);
+    if (event->check_and_clear()) {
+      if (this->service_is_ready()) {
+        return true;
+      }
+    }
+    // server is not ready, loop if there is time left
+    time_to_wait = timeout - (std::chrono::steady_clock::now() - start);
+  } while (timeout < std::chrono::nanoseconds(0) || time_to_wait > std::chrono::nanoseconds(0));
```

Comment:
That's a good question... Currently I think it will not wake up, because neither `node_->wait_for_graph_change` nor this loop will consider `rclcpp::ok` or be notified by the shutdown call or signal processor.

What do you think the behavior of this function should be in that case? I'll need to adjust the documentation to be something like "returns true if graph changed or false if the timeout expired or rclcpp::shutdown occurred". But then you can't easily tell which happened timeout or shutdown, though I guess if you check `rclcpp::ok` after getting `false` back from this function and be reasonably sure that was the cause or not.